tests/sentry/backup/test_correctness.py

```diff
@@ -0,0 +1,179 @@
+    exp_date_updated = parser.parse(expect["fields"]["date_updated"])
+    act_date_updated = parser.parse(actual["fields"]["date_updated"])
+    if not act_date_updated >= exp_date_updated:
+        return f"{expect['fields']['date_updated']} was not >= {actual['fields']['date_updated']}"
+    else:
+        expect["fields"]["date_updated"] = "__COMPARATOR_DATE_UPDATED__"
+        actual["fields"]["date_updated"] = "__COMPARATOR_DATE_UPDATED__"
+
+
+REPO_PATH = Path(os.path.dirname(os.path.realpath("__file__")))
+FIXTURE_PATH = REPO_PATH / "fixtures/backup"
+EXPORT_INDENTATION = 2
+COMPARATORS: dict[ModelName | None, list[JSONMutatingComparator]] = {
+    None: [],
+    ModelName("sentry.userrole"): [comparator_date_updated],
+    ModelName("sentry.userroleuser"): [comparator_date_updated],
+}
+
+
+def validate(expect: JSONData, actual: JSONData) -> ComparatorFindings:
+    """Ensures that originally imported data correctly matches actual outputted data, and produces a list of reasons why not when it doesn't"""
+
+    findings = ComparatorFindings([])
+    exp_models = {}
+    act_models = {}
+    for model in expect:
+        id = InstanceID(ModelName(model["model"]), PositiveInt(model["pk"]))
+        exp_models[id] = model
+
+    # Ensure that the actual JSON contains no duplicates - we assume that the expected JSON did not.
+    for model in actual:
+        id = InstanceID(ModelName(model["model"]), PositiveInt(model["pk"]))
+        if id in act_models:
+            findings.append(ComparatorFinding(ComparatorName("duplicate_entry"), id))
+        else:
+            act_models[id] = model
+
+    # Report unexpected and missing entries in the actual JSON.
+    act_ids = list(act_models)
+    exp_ids = list(exp_models)
+    for id in act_ids:
+        if id not in exp_ids:
+            del act_models[id]
+            findings.append(ComparatorFinding(ComparatorName("unexpected_entry"), id))
+    for id in exp_ids:
+        if id not in act_ids:
+            del exp_models[id]
+            findings.append(ComparatorFinding(ComparatorName("missing_entry"), id))
+
+    # We only perform custom comparisons and JSON diffs on non-duplicate entries that exist in both
+    # outputs.
+    for id, act in act_models.items():
+        exp = exp_models[id]
+
+        # Try global comparators and comparators applicable for this specific model.
+        for cmp in COMPARATORS[None]:
+            res = cmp(id, exp, act)
+            if res:
+                findings.append(ComparatorFinding(ComparatorName(cmp.__name__), id, res))
+        if id.model in COMPARATORS:
+            for cmp in COMPARATORS[id.model]:
+                res = cmp(id, exp, act)
+                if res:
+                    findings.append(ComparatorFinding(ComparatorName(cmp.__name__), id, res))
+
+        # Finally, perform a diff on the remaining JSON.
+        json_diff = diff(exp["fields"], act["fields"], syntax="explicit", marshal=True)
```

Comment:
ah here is the field comparison, what's the reason in having this here? is it because the `fields` schema can be so different based on the model?

Reply:
Yep, fields are different for each model, so we just do a simple text diff, modulo the funky fields that need special comparators.